src/StreamJsonRpc.Tests/JsonRpcClient20InteropTests.cs

```diff
@@ -149,6 +150,147 @@ public async Task InvokeWithParameterPassedAsObjectAsync_NoParameter()
+        {
+            string content = "{ \"jsonrpc\": \"2.0\", \"method\": \"$/progress\", \"params\": { \"token\": " + progressID + ", \"value\": " + i + ", } ,}";
+            JObject json = JObject.Parse(content);
+
+            this.Send(json);
+
+            sum2 += i;
+        }
+
+        System.Threading.Thread.Sleep(1000);
+        Assert.Equal(sum2, sum);
+
+        this.Send(new
+        {
+            jsonrpc = "2.0",
+            id = request["id"].Value<long>(),
+            result = sum,
+        });
+
+        int result = await invokeTask;
+        Assert.Equal(sum2, result);
+    }
+
+    [Fact]
+    public async Task InvokeWithProgressParameter_NoMemoryLeakConfirm()
+    {
+        WeakReference weakRef = await this.InvokeWithProgressParameter_NoMemoryLeakConfirm_Helper();
+        GC.Collect();
+        Assert.False(weakRef.IsAlive);
+    }
+
+    [Fact]
+    public async Task InvokeWithProgressParameterAsObject_CheckProgressObjectIsReplacedForId()
+    {
+        ProgressWithCompletion<int> progress = new ProgressWithCompletion<int>(report => { });
+
+        Task<int> invokeTask = this.clientRpc.InvokeWithParameterObjectAsync<int>("test", new { Bar = "value", Progress = progress });
+
+        JToken request = await this.ReceiveAsync();
+
+        Assert.Equal(JTokenType.Object, request["params"].Type);
+        Assert.Equal("value", request["params"]["Bar"].ToString());
+        Assert.Equal(JTokenType.Integer, request["params"]["Progress"].Type);
+    }
+
+    [Fact]
+    public async Task InvokeWithProgressParameterAsArray_CheckProgressObjectIsReplacedForId()
+    {
+        ProgressWithCompletion<int> progress = new ProgressWithCompletion<int>(report => { });
+
+        Task<int> invokeTask = this.clientRpc.InvokeAsync<int>("test", new object[] { "value", progress });
+
+        JToken request = await this.ReceiveAsync();
+
+        Assert.Equal(JTokenType.Array, request["params"].Type);
+        Assert.Equal("value", request["params"][0].ToString());
+        Assert.Equal(JTokenType.Integer, request["params"][1].Type);
+    }
+
+    [Fact]
+    public async Task InvokeWithProgressParameter_IncrementalId()
+    {
+        ProgressWithCompletion<int> progress = new ProgressWithCompletion<int>(report => { });
+
+        Task<int> invokeTask = this.clientRpc.InvokeAsync<int>("test", new object[] { "value", progress });
+        JToken request = await this.ReceiveAsync();
+
+        long firstId = request["params"][1].Value<long>();
+
+        invokeTask = this.clientRpc.InvokeAsync<int>("test", new object[] { "value", progress });
+        request = await this.ReceiveAsync();
+
+        Assert.Equal(firstId + 1, request["params"][1].Value<long>());
+
+        invokeTask = this.clientRpc.InvokeAsync<int>("test", new object[] { "value", progress });
+        request = await this.ReceiveAsync();
+
+        Assert.Equal(firstId + 2, request["params"][1].Value<long>());
+    }
+
     [Fact]
```

Comment:
This test is interesting in that it verifies that each request's progress gets a distinct token. But it goes about it by focusing on implementation details. Instead, can we simply create more than one concurrent request and report progress on each and verify that the right client `Progress<T>` object handles each of our $/progress messages?

Reply:
Also, this doesn't feel like it needs to be an interop test. Can we get better coverage for both JSON and MessagePack by moving to JsonRpcTests?

---
In reply to: [303656982](https://github.com/milopezc/vs-streamjsonrpc/pull/1#discussion_r303656982) [](ancestors = 303656982)